pgmpy/inference/mplp.py

```diff
@@ -0,0 +1,752 @@
+        self.last_obj = 1e40
+        self.m_intersect_map = []
+        for i in range(len(self.factors)):
+            self.objective.append([np.log(y) for y in self.factor_values[i]])
+            if len(self.factors[i]) > 1:
+                self.m_region_lambdas.append(self.objective[i])
+                curr_region = Region(self.factors[i], self.factors[:i + 1], self.factors[i], self.cardinality, i)
+                self.m_all_regions.append(curr_region)
+            if len(self.factors[i]) == 2:
+                self.m_intersect_map.append((self.factors[i], i))
+
+            # Initialise output vector
+            self.m_decoded_res = [0] * len(self.cardinality)
+            self.m_best_decoded_res = [0] * len(self.cardinality)
+
+    def _get_flat_ind(self, base_ind, base_card):
+        """
+        Converts:
+        base_ind=[1 0 0] to 4 for base_cardinality=[2 2 2]
+        base_ind=[1 0 0] to 6 for base_cardinality=[4 3 2]
+        """
+        return [list(i) for i in itertools.product(*[list(range(k)) for k in base_card])].index(base_ind)
+
+    def _integer_val(self):
+        """
+        The original factor potentials gives us one of the integer solutions.
+        """
+        int_val = 0
+        for i in range(len(self.m_all_regions)):
+            tmpvec = []
+            base_sizes = []
+            for j in self.m_all_regions[i].rvariables_inds:
+                tmpvec.append(self.m_decoded_res[j])
+                base_sizes.append(self.m_var_sizes[j])
+            int_val += self.m_region_lambdas[i][self._get_flat_ind(tmpvec, base_sizes)]
+        return int_val
+
+    def _local_decode(self):
+        """
+        Finds the best assignment from a set of assignment by locally maximizing the single node beliefs
+        """
+        obj = 0
+        for i in range(len(self.objective)):
+            obj += max(self.objective[i])
+            index_of_maximum = self.objective[i].index(max(self.objective[i]))
+            if len(self.m_all_intersect[i]) == 1:
+                self.m_decoded_res[self.m_all_intersect[i][0]] = index_of_maximum
+
+        # update the m_best_val if it is smaller than int_val
+        int_val = self._integer_val()
+        print('int_val: "{}"'.format(int_val))
+        if int_val > self.m_best_val:
+            self.m_best_decoded_res = self.m_decoded_res
+            self.m_best_val = int_val
+        return obj
+
+    def run_mplp(self, niter, obj_del_thr, int_gap_thr):
+        """
+        Perform the GMPLP updates.
+        """
+        for i in range(niter):
+            # For each of the 24 regions
+            for r in self.m_all_regions:
+                self.objective = self._update_msgs(self.objective, r)
+            obj = self._local_decode()
+            obj_del = self.last_obj - obj
+            self.last_obj = obj
+            int_gap = obj - self.m_best_val
+
+            print("m_best_val: {}".format(self.m_best_val))
```

Comment:
Don't put ```print``` statements in the final code. You can use it while debugging but make sure to remove them while sending a PR.